applications/nvidia_nim/nvidia_nim_imaging/app.py

```diff
@@ -0,0 +1,216 @@
+    "prompts": {"classes": ["liver", "spleen"]},
+}
+
+
+class PrintMessageOp(Operator):
+    def setup(self, spec: OperatorSpec):
+        spec.input("data")
+
+    def compute(self, op_input, op_output, context):
+        data = op_input.receive("data")
+
+        if data:
+            print(f"File saved in: {data}")
+
+
+class MessageBody:
+    def __init__(self, model_params: Dict, user_input: str):
+        self.model_params = model_params
+        self.user_input = user_input
+
+    def to_json(self):
+        return json.dumps(self.__dict__)
+
+    @classmethod
+    def from_json(cls, json_text):
+        return MessageBody(**json.loads(json_text))
+
+
+def get_api_key(app):
+    api_key = app.kwargs("nim")["api_key"]
+    if not api_key:
+        api_key = os.getenv("API_KEY", None)
+
+    if not api_key:
+        logger.warning("Setting up connection without an API key.")
+        logger.warning(
+            "Set 'api-key' in the nvidia_nim.yaml config file or set the environment variable 'API_KEY'."
+        )
+        print("")
+    return api_key
+
+
+class NIMOperator(Operator):
+    def __init__(
+        self,
+        fragment,
+        *args,
+        name,
+        base_url=None,
+        api_key=None,
+        **kwargs,
+    ):
+        self.base_url = base_url
+        self.api_key = api_key
+        self.model_params = dict(kwargs)
+
+        if self.api_key is None:
+            self.api_key = get_api_key(fragment)
+
+        # Need to call the base class constructor last
+        super().__init__(fragment, name=name)
+
+    def setup(self, spec: OperatorSpec):
+        spec.output("file_name")
+
+    def compute(self, op_input, op_output, context):
+        headers = {
+            "Authorization": f"Bearer {self.api_key}",
+        }
+
+        session = requests.Session()
+        response = session.post(self.base_url, headers=headers, json=payload)
+
+        response.raise_for_status()
+        op_output.emit(response.content, "file_name")
+
+
+class NimImaging(Application):
+    def __init__(self):
+        super().__init__()
+
+    def compose(self):
+        nim_op = NIMOperator(
+            self,
+            CountCondition(self, count=1),
+            name="nim",
+            **self.kwargs("nim"),
+        )
+
+        unzip_op = UnzipOp(
+            self, CountCondition(self, count=1), name="unzip", filter="*.response", output_path="."
+        )
+        print_files_op = PrintMessageOp(self, CountCondition(self, count=1), name="print_files")
+
+        self.add_flow(nim_op, unzip_op, {("file_name", "zip_file_bytes")})
+        self.add_flow(unzip_op, print_files_op, {("matching_files", "data")})
+
+    def set_nifti_file_path(self, file):
+        self._nifti_file_path = file
+        logger.info(f"NIFTI file: {self._nifti_file_path}")
+
+
+def _download_dataset(api_key, validate_file_checksum):
+    logger.info("Downloading sample image data from NVIDIA NGC...")
```

Comment:
I'm not clear, could you please clarify what is the difference between this dataset downloaded at init and the files fetched by `NIMOperator` at app runtime?

Reply:
This downloads the input dataset.  The operator returns the result from the Vista3D API call.

Reply:
Understood!